pkg/config/v3alpha/config.go

```diff
@@ -0,0 +1,340 @@
+}
+
+// PluginConfigs holds a set of arbitrary plugin configuration objects mapped by plugin key.
+// TODO: do not export this once internalconfig has merged with config
+type PluginConfigs map[string]pluginConfig
+
+// pluginConfig is an arbitrary plugin configuration object.
+type pluginConfig interface{}
+
+// New returns a new config.Config
+func New() config.Config {
+	return &cfg{Version: Version}
+}
+
+func init() {
+	config.Register(Version, New)
+}
+
+// GetVersion implements config.Config
+func (c cfg) GetVersion() config.Version {
+	return c.Version
+}
+
+// GetDomain implements config.Config
+func (c cfg) GetDomain() string {
+	return c.Domain
+}
+
+// SetDomain implements config.Config
+func (c *cfg) SetDomain(domain string) error {
+	c.Domain = domain
+	return nil
+}
+
+// GetRepository implements config.Config
+func (c cfg) GetRepository() string {
+	return c.Repository
+}
+
+// SetRepository implements config.Config
+func (c *cfg) SetRepository(repository string) error {
+	c.Repository = repository
+	return nil
+}
+
+// GetName implements config.Config
+func (c cfg) GetName() string {
+	return c.Name
+}
+
+// SetName implements config.Config
+func (c *cfg) SetName(name string) error {
+	c.Name = name
+	return nil
+}
+
+// GetLayout implements config.Config
+func (c cfg) GetLayout() string {
+	return c.Layout
+}
+
+// SetLayout implements config.Config
+func (c *cfg) SetLayout(layout string) error {
+	c.Layout = layout
+	return nil
+}
+
+// IsMultiGroup implements config.Config
+func (c cfg) IsMultiGroup() bool {
+	return c.MultiGroup
+}
+
+// SetMultiGroup implements config.Config
+func (c *cfg) SetMultiGroup() error {
+	c.MultiGroup = true
+	return nil
+}
+
+// ClearMultiGroup implements config.Config
+func (c *cfg) ClearMultiGroup() error {
+	c.MultiGroup = false
+	return nil
+}
+
+// IsComponentConfig implements config.Config
+func (c cfg) IsComponentConfig() bool {
+	return c.ComponentConfig
+}
+
+// SetComponentConfig implements config.Config
+func (c *cfg) SetComponentConfig() error {
+	c.ComponentConfig = true
+	return nil
+}
+
+// ClearComponentConfig implements config.Config
+func (c *cfg) ClearComponentConfig() error {
+	c.ComponentConfig = false
+	return nil
+}
+
+// Resources implements config.Config
+func (c cfg) ResourcesLength() int {
+	return len(c.Resources)
+}
+
+// HasResource implements config.Config
+func (c cfg) HasResource(gvk resource.GVK) bool {
+	for _, res := range c.Resources {
+		if gvk.IsEqualTo(res.GVK()) {
+			return true
+		}
+	}
+
+	return false
+}
+
+// GetResource implements config.Config
+func (c cfg) GetResource(gvk resource.GVK) (resource.Resource, error) {
+	for _, res := range c.Resources {
+		if gvk.IsEqualTo(res.GVK()) {
+			return res, nil
+		}
+	}
+
+	return resource.Resource{}, config.UnknownResource{GVK: gvk}
+}
+
+// AddResource implements config.Config
+func (c *cfg) AddResource(res resource.Resource) error {
+	// TODO: remove when these fields need to be tracked
+	res.Domain = ""
```

Comment:
IMO we should NOT store the Domain or have its attr in the resource. 
The domain is == for the full config project and is NOT specific for a resource. See: https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v3/PROJECT#L1

Reply:
I completely disagree. In k8s, a resource is uniquely identified by its GVK (Group, Version and Kind) where the group has a domain. See: https://github.com/kubernetes/kubernetes/blob/bb376f161640529f600203faee4cd0de8ec16778/pkg/apis/admission/register.go#L25

To simplify the process for the user, in kubebuilder we have decided to allow a project-wide domain to be configured for every group, so that the user can write `--group group` instead of `--group group.domain.com` every time he wants to create an API or a webhook. To keep the UX in sync with the `Resource` model, `Group` has been kept to refer to the value of the `--group` flag, and `QualifiedGroup` to refer to the group as understood by k8s (i.e., including the domain). This proves that the domain is part of the resource definition.

Additionally, we are currently supporting core resources which do not use the project-wide domain. See https://github.com/kubernetes-sigs/kubebuilder/blob/26d7b42314dc8df4369df3594329c3d2401793d2/testdata/project-v3-multigroup/controllers/crew/captain_controller.go#L37 and https://github.com/kubernetes-sigs/kubebuilder/blob/26d7b42314dc8df4369df3594329c3d2401793d2/testdata/project-v3-multigroup/controllers/apps/pod_controller.go#L35 to see how the domain is needed per resource. This proves the need of  of this field in order to scaffold the files. (In the future we also plan for additional support for external resources which will also have different domains.)

P.S.: As of this PR, the config file hasn't been modified to include this field.